refactor: Improve DevX

Karrq opened the pull request:
The aim of this PR is to improve the DevX by providing a more streamlined set of scripts to setup and start testing the environment locally, but also aid during first checkout, by fetching all required external dependencies, not otherwise present in the repo.

NOTE: Originally the PR moved rabby from the root to `deps/rabby` for consistency, but that exploded the number of changed files. So this PR still has the assumption that rabby lives in `deps/rabby`, but the rabby fork will be moved in a separate PR.

TODO:
- [X] External dependency management
- [X] Migrate gateway config and start
- [X] Migrate main node config and start (with existing chain)
- [X] Migrate L2 chain deployment
- [x] Migrate common operations (start/stop, logs, ...)
- [x] Migrate testing recipes
- [x] Update README

Karrq committed:
chore: move rabby under deps/rabby

Karrq committed:
feat(deps): fetch external deps

Karrq committed:
refactor: based/docker

Karrq committed:
fix(metrics-exporter): remove reth reference

fix(build:docker): quiet invocations

Karrq committed:
feat(docs): Justfile

feat: justfile doctor (system dependencies check)

Karrq committed:
feat(gateway): justfile

Karrq committed:
refactor: avoid import, invoke instead

Karrq committed:
feat: main node cfg

refactor(gateway): external env config

Karrq committed:
feat: deploy.just

Karrq committed:
chore: rename task

Karrq committed:
fix(main-node): typos

feat(spamoor): migrate spamoor start script

Karrq committed:
feat: logs

pamungkaski commented:
@claude plz review

Karrq committed:
feat(just): tests

Karrq committed:
chore: restore rabby at ./rabby

Karrq committed:
fix: gateway -> follower-node

Karrq committed:
feat: remaining tests

Karrq committed:
docs: update README

feat: overseer
refactor: other-logs -> logs
feat: start/stop/logs component

Karrq committed:
Merge remote-tracking branch 'origin/develop' into refactor/devex

Karrq commented:
There are some things I'd like to clarify to improve this further:
1. `BOP_EL_PORT`, `OP_EL_PORT`, `BOP_NODE_PORT` and `SEQ` seem to be used but I'm unable to determine what their values should be
2. `fix-compose` recipe is used but I can't find its definition
3. `spamoor-daemon` folder is used but can't find it in the repo
4. how are gateway-spam and gateway-bench supposed to be used?

Karrq committed:
fix: deploy-chain

fix: typos

Karrq committed:
fix: logs

fix: IPv6 public logs
fix: more typos
fix: register-gateway
refactor: scripts/overseer -> based/overseer

Karrq committed:
fix: register-gateway

fix: public-ip
perf: public-ip

Karrq committed:
fix(test): continue lines

BrycePy commented on based/scripts/deploy.just:
```diff
@@ -0,0 +1,120 @@
+    fi
+
+    if [ -d $cfg ]; then
+        echo "❌ Seems like information of a previous chain is already present. Please remove $cfg to deploy a new one."
```
Shouldn't continue if this echo is called

BrycePy commented on based/main-node.just:
```diff
@@ -0,0 +1,182 @@
+     
+    # Generate JWT if missing
+    if [ ! -f $MAIN_NODE_DATA/jwt ]; then
+      openssl rand -hex 32 | tr -d '\n' | sed 's/^/0x/' > $MAIN_NODE_DATA/jwt
```
this is in the wrong path. should be $MAIN_NODE_DATA/config/jwt

BrycePy commented on based/follower-node.just:
```diff
@@ -0,0 +1,178 @@
+     
+    # Generate JWT if missing
+    if [ ! -f $FOLLOWER_NODE_DATA/jwt ]; then
+      openssl rand -hex 32 | tr -d '\n' | sed 's/^/0x/'  > $FOLLOWER_NODE_DATA/jwt
```
wrong dir here too. should be $FOLLOWER_NODE_DATA/config/jwt

BrycePy commented on based/follower-node.just:
```diff
@@ -0,0 +1,178 @@
+    docker logs based-gateway --tail 100 -f
+
+logs $dotenv=join(FOLLOWER_NODE_DATA, ".env"):
+    docker compose -f $(dirname $FOLLOWER_NODE_DATA)/compose.yml logs --tail 100 -f
```
should be `docker compose -f $(dirname $dotenv)/compose.yml logs --tail 100 -f`

BrycePy commented on based/follower-node.just:
```diff
@@ -0,0 +1,178 @@
+    # Start services
+    echo "Starting gateway and monitoring services..."
+    docker compose -f $FOLLOWER_NODE_DATA/compose.yml up -d
+    {{parent}} monitoring start
```
Could you make this optional? maybe a new function start-dev that doesnt start the monitoring?

Karrq committed:
fix: deploy-chain populate-intent call

Karrq committed:
fix: PR review

feat: start without monitoring
fix: jwt location
fix: exit after previous chain deployment is detected

Karrq committed:
refactor: deps module as invocation

Karrq committed:
fix: _populate-intent location

BrycePy committed:
working reset-and-start-full-stack-local

Karrq committed:
feat: wallet backup

BrycePy committed:
fix

Karrq committed:
fix: wallet backup

refactor: reset-and-start-full-stack-local

BrycePy committed:
increase overseer start delay

Karrq committed:
fix: wallet typo

BrycePy committed:
update example port

Karrq committed:
feat: reset follower/main node

fix(follower-node): portal RPC queries
fix(portal): error on null/empty RPC result
chore: op-deployer containers autoremove

Karrq committed:
feat: ci script & deployed chain assets

Karrq committed:
feat(script): peering

Karrq committed:
fix(just): peering

Karrq committed:
feat: ci script from main justfile

fix(ci): prepare checks full config

BrycePy committed:
fix + peering

Karrq committed:
fix: peering force local enode

fix(test): defaults for env variables

Karrq committed:
feat: .env.example

feat(ci): restore wallets from backup

Karrq committed:
fix: local address for multiaddress

Karrq committed:
fix(make): DisputeGameFactoryProxy

fix(make): correct defaults for L1 RPCs
fix(deploy): set chain cfg folder permissions

Karrq committed:
feat(compose): prod/dev setup

Karrq committed:
feat(ci): integration test

Karrq committed:
fix(ci): export correct image

Karrq committed:
feat: parallel build

Karrq committed:
fix(dev): container-name

Karrq committed:
fix(ci): run against develop branch too

Karrq committed:
fix: parallel deps build

Karrq committed:
fix(ci): checkout repo

Karrq committed:
fix(ci): use .env.example

Karrq committed:
fix(ci): remove keys from .env.example

fix(build): parallel thru just, hash recipe

Karrq committed:
fix(ci): ensure LOCAL_DATA

Karrq committed:
fix(ci): upload hidden .local

fix(ci): more portable test ports

Karrq committed:
fix(ci): ignore .local/data

Karrq committed:
fix(ci): ignore config artifact

Karrq committed:
fix(ci): BASED_OP_GETH_PORT from .env

fix: configurable PORTAL_PORT

Karrq committed:
fix(scripts): intent permission

fix(scripts): display existing wallet address with ensure

Karrq committed:
feat(ci): upgrade chain config

Karrq committed:
feat(ci): workflow_call only

Karrq committed:
feat(ci): integration triggers

Karrq committed:
fix(ci): config-with-deploy

Karrq committed:
feat(ci): layer caching

BrycePy commented:
Commit dcf5bb2ddc9b4ed178620fbe6766a7230501ac0a seems to break the `just reset-and-start-full-stack-local` command. I'm getting the following errors:

```
Resetting configuration and deploying new L2 from scratch
just -f /home/su/based-op/Justfile main-node reset
just -f based/main-node.just reset
cat: /home/su/based-op/.local/.wallet/batcher/key: No such file or directory
error: Recipe `key` failed on line 68 with exit code 1
error: Call to function `shell` failed: Process exited with status code 1
  ——▶ main-node.just:20:48
   │
20 │ export OP_BATCHER_KEY := env("OP_BATCHER_KEY", shell(wallet + " key batcher"))
   │                                                ^^^^^
error: Recipe `main-node` failed on line 66 with exit code 1
error: Recipe `reset` failed on line 96 with exit code 1
just -f based/main-node.just config-with-deploy
cat: /home/su/based-op/.local/.wallet/batcher/key: No such file or directory
error: Recipe `key` failed on line 68 with exit code 1
error: Call to function `shell` failed: Process exited with status code 1
  ——▶ main-node.just:20:48
   │
20 │ export OP_BATCHER_KEY := env("OP_BATCHER_KEY", shell(wallet + " key batcher"))
   │                                                ^^^^^
error: Recipe `main-node` failed on line 66 with exit code 1
error: Recipe `reset-and-start-full-stack-local` failed with exit code 1
```

Karrq committed:
fix: more robust wallet keys
